zkfacade/src/main/java/com/yahoo/vespa/curator/Lock.java

```diff
@@ -15,31 +16,46 @@
  */
 public class Lock implements Mutex {
 
+    private final ReentrantLock lock;
     private final InterProcessLock mutex;
     private final String lockPath;
 
     public Lock(String lockPath, Curator curator) {
         this.lockPath = lockPath;
+        this.lock = new ReentrantLock(true);
         mutex = curator.createMutex(lockPath);
     }
 
     /** Take the lock with the given timeout. This may be called multiple times from the same thread - each matched by a close */
     public void acquire(Duration timeout) throws UncheckedTimeoutException {
-        boolean acquired;
+        boolean acquired = false;
         try {
             acquired = mutex.acquire(timeout.toMillis(), TimeUnit.MILLISECONDS);
+            lock.tryLock(); // Should be available to only this thread, while holding the above mutex.
         }
         catch (Exception e) {
+            if (acquired) release();
             throw new RuntimeException("Exception acquiring lock '" + lockPath + "'", e);
         }
 
-        if (! acquired)
+        if ( ! lock.isHeldByCurrentThread()) {
```

Comment:
I found the previous version of this code easer to read, probably because of the implicit non-obvious assumption that tryLock() is assumed to always return true, and then relying on that here.

Reply:
Right. Fixed. PTAL.